src/unix_term.rs

```diff
@@ -360,3 +361,289 @@ pub fn wants_emoji() -> bool {
 pub fn set_title<T: Display>(title: T) {
     print!("\x1b]0;{}\x07", title);
 }
+
+fn with_raw_terminal<R>(f: impl FnOnce(&mut fs::File) -> R) -> io::Result<R> {
+    // We need a custom drop implementation for File,
+    // so that the fd for stdin does not get closed
+    enum CustomDropFile {
+        CloseFd(Option<fs::File>),
+        NotCloseFd(Option<fs::File>),
+    }
+
+    impl Drop for CustomDropFile {
+        fn drop(&mut self) {
+            match self {
+                CustomDropFile::CloseFd(_) => {}
+                CustomDropFile::NotCloseFd(inner) => {
+                    if let Some(file) = inner.take() {
+                        file.into_raw_fd();
+                    }
+                }
+            }
+        }
+    }
+
+    let (mut tty_handle, tty_fd) = if unsafe { libc::isatty(libc::STDIN_FILENO) } == 1 {
+        (
+            CustomDropFile::NotCloseFd(Some(unsafe { fs::File::from_raw_fd(libc::STDIN_FILENO) })),
+            libc::STDIN_FILENO,
+        )
+    } else {
+        let handle = fs::OpenOptions::new()
+            .read(true)
+            .write(true)
+            .open("/dev/tty")?;
+        let fd = handle.as_raw_fd();
+        (CustomDropFile::CloseFd(Some(handle)), fd)
+    };
+
+    // Get current mode
+    let mut termios = mem::MaybeUninit::uninit();
+    c_result(|| unsafe { libc::tcgetattr(tty_fd, termios.as_mut_ptr()) })?;
+
+    let mut termios = unsafe { termios.assume_init() };
+    let old_iflag = termios.c_iflag;
+    let old_oflag = termios.c_oflag;
+    let old_cflag = termios.c_cflag;
+    let old_lflag = termios.c_lflag;
+
+    // Go into raw mode
+    unsafe { libc::cfmakeraw(&mut termios) };
+    if old_lflag & libc::ISIG != 0 {
+        // Re-enable INTR, QUIT, SUSP, DSUSP, if it was activated before
+        termios.c_lflag |= libc::ISIG;
+    }
+    c_result(|| unsafe { libc::tcsetattr(tty_fd, libc::TCSADRAIN, &termios) })?;
+
+    let result = match &mut tty_handle {
+        CustomDropFile::CloseFd(Some(handle)) => f(handle),
+        CustomDropFile::NotCloseFd(Some(handle)) => f(handle),
+        _ => unreachable!(),
+    };
+
+    // Reset to previous mode
+    termios.c_iflag = old_iflag;
+    termios.c_oflag = old_oflag;
+    termios.c_cflag = old_cflag;
+    termios.c_lflag = old_lflag;
+    c_result(|| unsafe { libc::tcsetattr(tty_fd, libc::TCSADRAIN, &termios) })?;
+
+    Ok(result)
+}
+
+pub fn supports_synchronized_output() -> bool {
+    *sync_output::SUPPORTS_SYNCHRONIZED_OUTPUT
+}
+
+/// Specification: https://gist.github.com/christianparpart/d8a62cc1ab659194337d73e399004036
+mod sync_output {
```

Comment:
I think Rust style would suggest not abbreviating `synchronized_` to `sync_` for the module name.